server-spi-private/src/main/java/org/keycloak/validate/ValidatorConfig.java

```diff
@@ -0,0 +1,184 @@
+        return getStringOrDefault(key, null);
+    }
+
+    public String getStringOrDefault(String key, String defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof String) {
+            return (String) value;
+        }
+        return defaultValue;
+    }
+
+    public Integer getIntOrDefault(String key) {
+        return getIntOrDefault(key, null);
+    }
+
+    public Integer getIntOrDefault(String key, Integer defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof Integer) {
+            return (Integer) value;
+        } else if (value instanceof Number) {
+            return ((Number) value).intValue();
+        } else if (value instanceof String) {
+            return Integer.parseInt((String) value);
+        }
+        return defaultValue;
+    }
+
+    public Double getDoubleOrDefault(String key) {
+        return getDoubleOrDefault(key, null);
+    }
+
+    public Double getDoubleOrDefault(String key, Double defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof Double) {
+            return (Double) value;
+        } else if (value instanceof Number) {
+            return ((Number) value).doubleValue();
+        } else if (value instanceof String) {
+            return Double.parseDouble((String) value);
+        }
+        return defaultValue;
+    }
+
+    public Boolean getBooleanOrDefault(String key) {
+        return getBooleanOrDefault(key, null);
+    }
+
+    public Boolean getBooleanOrDefault(String key, Boolean defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof Boolean) {
+            return (Boolean) value;
+        } else if (value instanceof String) {
+            return Boolean.parseBoolean((String) value);
+        }
+        return defaultValue;
+    }
+
+    public Set<String> getStringSetOrDefault(String key) {
+        return getStringSetOrDefault(key, null);
+    }
+
+    public Set<String> getStringSetOrDefault(String key, Set<String> defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof Set) {
+            return (Set<String>) value;
+        }
+        return defaultValue;
+    }
+
+    public List<String> getStringListOrDefault(String key) {
+        return getStringListOrDefault(key, null);
+    }
+
+    public List<String> getStringListOrDefault(String key, List<String> defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof List) {
+            return (List<String>) value;
+        }
+        return defaultValue;
+    }
+
+    public Pattern getPattern(String key) {
+        return getPatternOrDefault(key, null);
+    }
+
+    public Pattern getPatternOrDefault(String key, Pattern defaultValue) {
+        Object value = config.get(key);
+        if (value instanceof Pattern) {
+            return (Pattern) value;
+        } else if (value instanceof String) {
+            return Pattern.compile((String) value);
```

Comment:
Should we store the compiled pattern for the next time?

Reply:
I'd delegate that to the ValidatorConfig. But yes, the pattern could be cached.